### DISCUSSION TITLE
Execute ruff only on staged changes

### DISCUSSION THREAD

**Discussion by lexicalunit** [astral-sh/ruff#4049] [Q&A] [answered]
We want to start using ruff to prevent bad code from making it into future PRs while we work on fixing the issues that `ruff` surfaces about the entire codebase over time.

Is there a way to only run ruff on git staged changes rather than the entire source code?

**Comment by dhruvmanila** [comment]
I think you could use `pre-commit` (https://pre-commit.com/) to do it, but it'll still be scoped to the entire contents of all the staged files.

**Reply by charliermarsh** [reply]
I think @dhruvmanila is right that you could use `pre-commit` if you're only interested in running on changed files.

**Comment by evanrittenhouse** [comment]
I did find this [snippet](https://gist.github.com/dahjelle/8ddedf0aebd488208a9a7c829f19b9e8) which may be able to help out. On mobile so not able to try it though - you'd probably need to fiddle around to change the command from `eslint` to `ruff`

**Comment by charliermarsh** [comment, marked as answer]
Ah yeah, unfortunately we don't support this right now, though it's come up a few times in different forms: #1149,  #1931, #2696. I've generally been a bit hesitant to support it since it seems quite hard to get right.

**Reply by jstriebel** [reply]
Probably you are already aware of it, but the [darker](https://github.com/akaihola/darker) tool does exactly that for other formatters and linters. (Edit: Sorry, just found https://github.com/charliermarsh/ruff/issues/2472 after posting this)

**Reply by itolosa** [reply]
I've tried to run ruff using darker like this:

```
darker -L "ruff check" . --check --diff
```

it seems to work fine @jstriebel

**Reply by OdysseasKr** [reply]
> I've tried to run ruff using darker like this:
> 
> ```
> darker -L "ruff check" . --check --diff
> ```
> 
> it seems to work fine @jstriebel


This seems to work great up until ruff==0.12.8. From 0.12.9 it no longer works unfortunately.

**Reply by charliermarsh** [reply]
Please file a separate issue and be specific about whatever behavior change you're seeing. We prefer to avoid discussion on closed issues and questions, and we can't really help you if we don't know what problem you're running into.

**Comment by HealthyPear** [comment]
Not really straightforward as having a direct option flag from `ruff`, but what about simply using `git` functionalities?

`ruff check $(git diff --name-only --cached -- '*.py')`

You can also do a similar thing on a CI pipeline over files changed via commits with respect to the master branch, for example in the case of a Merge Request pipeline on GitLab ,

`ruff check $($(git diff --name-only $CI_MERGE_REQUEST_DIFF_BASE_SHA -- '*.py'))`

**Reply by codethief** [reply]
If I have staged only some (not all) changes via `git add -p`, won't this touch the files in the working tree (as opposed to those in the staging area)?

**Reply by HealthyPear** [reply]
Maybe yes, the command above gives a list of files that have been added to the staging area; if you stage only a patch, that still comes from a file, so using something based on git diff might always result in the filename

maybe for that specific use case there is a better option

**Comment by dcdanielca** [comment]
This works perfectly:

`git diff -u --relative --name-only | ruff check`

`git diff -u --relative --name-only | ruff format --check`

This commands you can run with pre-commit or lefthook. For example lefthook.yml config:
```
pre-push:
  parallel: true
pre-commit:
  commands:
    ruff-linter:
      run: git diff -u --relative --name-only | ruff check
    run-format:
      run: git diff -u --relative --name-only | ruff format --check
```

**Reply by codethief** [reply]
I don't think this works as expected when you stage only patches, see my [previous comment](https://github.com/astral-sh/ruff/discussions/4049#discussioncomment-8324334).

**Comment by paylogic-blaise** [comment]
It's possible to run `ruff check` on modified lines only, and ignore violations on existing lines using https://pypi.org/project/diff-cover/:

```
pip install diff-cover 
diff-quality --violations=ruff.check --compare-branch=origin/main
```

`diff-quality` uses its own custom format, not Ruff's, so this will not be useful for many use cases, but it may match @lexicalunit's.
We use it to block the introduction of `ruff check` violations on new code only in a CI/CD job.

**Reply by MichaReiser** [reply]
No, ruff doesn't support this on its own. See https://github.com/astral-sh/ruff/issues/1149

**Reply by paylogic-blaise** [reply]
I perhaps phrased it confusingly, it works with the `diff-quality` cli tool.

**Reply by OdysseasKr** [reply]
Another great suggestion but unfortunately, much like the `darker` suggestion, it broke with ruff 0.12.9
